### PR TITLE
Implement optional type parameters

### PR DESCRIPTION
**Pull request by dcodeIO**
This implements syntax like `someFunc<T, TIn=T>(...)`, with `TIn` being omittable, becoming the default type. Also touches all the `loadUnsafe` etc. stuff, renaming these to more macro-like style. Includes suggestions on how to proceed with `TypedArray<T,TIn>` as well, using macros. Wdyt?

### CODE REVIEW

File: src/util/collections.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+export function uniqueArray<V>(original: Array<V> | null = null): Array<V> {
+  if (original) {
+    let cloned = new Array<V>(original.length);
+    for (let i = 0, k = original.length; i < k; ++i) cloned[i] = original[i];
+    return cloned;
+  }
+  return new Array<V>();
```
Comment (MaxGraey) [inline src/util/collections.ts]: Fastest and simplest unique for array which I know is:
```js
function unique<T>(arr: Array<T>): Array<T> {
  return Array.from(new Set<T>(arr));
}
```
But it require iterators. And may be not perfect align to current stale of AS
```

File: src/util/collections.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+export function uniqueSet<V>(original: Set<V> | null = null): Set<V> {
+  if (original) {
+    let cloned = new Set<V>();
+    for (let v of original) cloned.add(v);
+    return cloned;
+  }
+  return new Set<V>();
+}
```
Comment (MaxGraey) [inline src/util/collections.ts]: Not sure this make sense. Any `Set` already contain only unique values

File: src/util/collections.ts (hunk 1 above)
Comment (MaxGraey) [reply src/util/collections.ts]: Ahh, I got what is that) May be better rename this to `makeArray` and `makeSet` for avoid confusion?

Comment (dcodeIO) [reply src/util/collections.ts]: Yeah, the code currently assumes that we'd at some point try to compile the compiler with asc, with some features not yet supported.

Comment (MaxGraey) [reply src/util/collections.ts]: I got it. But probably make sense rename it to `make*` instead `unique*` ?

Comment (dcodeIO) [reply src/util/collections.ts]: Fine with `make*` as well, going to change :)

**Comment by dcodeIO** [comment]
@MaxGraey anything else I should take a look at, or do you think this is good to merge?

**Comment by dcodeIO** [comment]
The issue there is that as soon as the offset becomes an argument (and this a local), it is no longer constant and can't be used as the constant `offset=` immediate of a load or store. Hence the two versions of the macro. Not sure if this can be worked around without introducing new syntax or something like that.

**Comment by MaxGraey** [comment]
Yeah, I forgot we coudn't verify constant before optimization. So in this case I do not see any obstacles to merge

**Comment by dcodeIO** [comment]
Just noticed that it isn't making use of the `offset=` immediate actually. So it should actually optimize, going to try.

**Comment by dcodeIO** [comment]
See the commit above. Looks like this only moves around a few locals in the optimized output, nothing major, and should be good to do.

**Comment by MaxGraey** [comment]
Now looks good =)

**Pull request merged**